### PR TITLE
refactor!: Define methods for `ape::as.phylo()` instead of own `as_phylo()` generic, remove deprecated `asPhylo()`

### LINKED ISSUE #713: why is as_phylo.igraphHRG not exported
I'm working on an automated script for #708 and noticed this.

If the function isn't exported, the lifecycle stuff isn't so important.

### PR DESCRIPTION
**Pull request by maelle**
Fix #713 

Not really breaking since it's still a soft deprecation.

### CODE REVIEW

File: NAMESPACE [hunk 1]
```diff
@@ -139,8 +141,8 @@ export(articulation.points)
 export(articulation_points)
 export(as.directed)
 export(as.igraph)
+export(as.phylo)
 export(as.undirected)
-export(asPhylo)
```
Comment (maelle) [inline NAMESPACE:143]: this one I really killed.

Comment (maelle) [reply NAMESPACE:143]: Which seems fine given the occurrences of the same name in https://github.com/search?q=org%3Acran+asPhylo+&type=code are not from this igraph method.

File: R/community.R [hunk 2]
```diff
@@ -696,15 +696,15 @@ as.hclust.communities <- function(x, hang = -1, use.modularity = FALSE,
 #' @rdname communities
 #' @family community
 #' @export
-as_phylo <- function(x, ...) {
-  UseMethod("as_phylo")
+as.phylo <- function(x, ...) {
```
Comment (krlmlr) [inline R/community.R:699]: There is already `ape::as.phylo()`, do we really need to do anything here?

File: R/community.R [hunk 3]
```diff
@@ -696,15 +696,15 @@ as.hclust.communities <- function(x, hang = -1, use.modularity = FALSE,
...
+  UseMethod("as.phylo")
 }
 
 #' @rdname communities
-#' @method as_phylo communities
+#' @method as.phylo communities
```
Comment (krlmlr) [inline R/community.R:704]: Do we need to use `s3_register()` in `.onLoad()` here?

Comment (maelle) [reply R/community.R:704]: I'm a bit confused (and not very experienced with S3). That's what you suggested and what I added in `R/aaa-utils.R`, was that wrong?

File: R/community.R (hunk 2 above)
Comment (maelle) [reply R/community.R:699]: Should this just be deleted? As noted in the other comment I'm now confused.

File: R/aaa-utils.R [hunk 4]
```diff
@@ -0,0 +1,3 @@
+.onLoad <- function(...) {
```
Comment (krlmlr) [inline R/aaa-utils.R:1]: Should we do it the rlang way? Should probably live in `R/zzz.R` then.

https://rlang.r-lib.org/reference/on_load.html?q=onload#comparison-with-onload-

File: R/community.R (hunk 2 above)
Comment (krlmlr) [reply R/community.R:699]: Yes, I guess we don't need it here.

File: R/community.R (hunk 3 above)
Comment (krlmlr) [reply R/community.R:704]: We don't need `#' @method`, I believe.

**Review by krlmlr — COMMENTED** [review]
Thanks, looks good!

**Review by krlmlr — APPROVED** [review]
Looks good, except our `as.phylo()` generic. Can you please fix and merge when CI passes?

**Comment by krlmlr** [comment]
Why do we need `UseMethod("ape::as.phylo")` ?

What happens if you remove the `as.phylo()` export, without replacement?

File: NAMESPACE [hunk 5]
```diff
@@ -141,7 +140,7 @@ export(articulation.points)
 export(articulation_points)
 export(as.directed)
 export(as.igraph)
-export(as.phylo)
+export(as.phylo.communities)
```
Comment (krlmlr) [inline NAMESPACE:143]: This needs to go to.

**Comment by krlmlr** [comment]
R CMD check errors:

```
  as_phylo.igraphHRG: no visible global function definition for
    ‘as.phylo’
  dendPlotPhylo: no visible global function definition for ‘as.phylo’
  hrgPlotPhylo: no visible global function definition for ‘as.phylo’
  Undefined global functions or variables:
    as.phylo
```

File: NAMESPACE (hunk 5 above)
Comment (krlmlr) [reply NAMESPACE:143]: - Remove the appropriate `#' @export` tag
- Ensure that `as.phylo.communities()` is registered on load (if ape is loaded) or as soon as ape is loaded, by calling `s3_register()` in `.onLoad()` or via `rlang::run_on_load()`

File: NAMESPACE [hunk 6]
```diff
@@ -30,6 +30,7 @@ S3method(as.matrix,igraph)
 S3method(as_ids,igraph.es)
 S3method(as_ids,igraph.vs)
 S3method(as_phylo,communities)
+S3method(as_phylo,igraphHRG)
```
Comment (krlmlr) [inline NAMESPACE:33]: I wonder why these entries still exist. Should they also be methods for the `ape::as.phylo()` generic and registered via `s3_register()` .

**Review by krlmlr — COMMENTED** [review]
For the CRAN failures, you need to consistently call `ape::as.phylo()` instead of `as.phylo()` .

**Comment by krlmlr** [comment]
Do you need more input here?

**Comment by maelle** [comment]
I'm sorry, I'm confused once again. I thought we needed to remove `as.phylo.communities` but it's only defined in this package so let me add it back...

**Comment by maelle** [comment]
@krlmlr I'm still not entirely sure about the changes, even if R CMD check seems happy, so would appreciate a review.

File: NAMESPACE (hunk 6 above)
Comment (krlmlr) [reply NAMESPACE:33]: I still think this needs to go. All the `as.phylo()` methods should be registered dynamically when/if ape is loaded.

File: R/as_phylo.R [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+#' @family community
+#' @export
+#' @param x object to be coerced
+#' @inheritParams ape::as.phylo
+as_phylo <- function(x, ...) {
+  lifecycle::deprecate_soft(
+    "1.5.0",
+    "ape::as.phylo()"
+  )
+  UseMethod("ape::as.phylo")
```
Comment (krlmlr) [inline R/as_phylo.R:17]: I'm not sure that this is going to work. This should check that ape is installed and forward to `ape::as.phylo()`, with a deprecation message. This is no longer a generic, we don't need `UseMethod()` here.

Or, alternatively, with or without deprecation message:

File: R/as_phylo.R [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+#' @rdname as_phylo
+#' @inheritParams ape::as.phylo
+#' @export
+as_phylo.igraphHRG <- function(x, ...) {
+  lifecycle::deprecate_soft(
+    "1.5.0",
+    "as.phylo()"
+  )
+  ape::as.phylo(x, ...)
+}
```
Comment (krlmlr) [inline R/as_phylo.R:41]: We don't need those if the generic is gone.

File: R/as_phylo.R [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+rlang::on_load(s3_register("ape::as.phylo", "igraphHRG"))
```
Comment (krlmlr) [inline R/as_phylo.R:42]: Should this live closer to the declaration of the method?

File: R/community.R [hunk 10]
```diff
@@ -96,7 +96,7 @@
 #' `as.hclust` is similar to [as.dendrogram()], but converts a
 #' hierarchical community structure to a `hclust` object.
 #'
-#' `as_phylo()` converts a hierarchical community structure to a `phylo`
+#' `as.phylo()` converts a hierarchical community structure to a `phylo`
```
Comment (krlmlr) [inline R/community.R:99]:
```suggestion
#' [ape::as.phylo()] converts a hierarchical community structure to a `phylo`
```

File: R/community.R [hunk 11]
```diff
@@ -96,7 +96,7 @@
 #' `as.hclust` is similar to [as.dendrogram()], but converts a
```
Comment (krlmlr) [inline R/community.R:96]:
```suggestion
#' [stats::as.hclust()] is similar to [as.dendrogram()], but converts a
```

File: R/as_phylo.R (hunk 7 above)
Comment (krlmlr) [inline R/as_phylo.R:17]:
```suggestion
  check_installed("ape", "for using `as.phylo()`")
  UseMethod("as.phylo")
```

File: NAMESPACE (hunk 6 above)
Comment (maelle) [reply NAMESPACE:33]: But if they don't exist at all how would that not break old  code?

File: R/as_phylo.R (hunk 9 above)
Comment (maelle) [reply R/as_phylo.R:43]: You mean `as.phylo.igraphHRG()` ?

**Review by krlmlr — CHANGES_REQUESTED** [review]
I can take a look here.

**Review by krlmlr — APPROVED** [review]

**Pull request merged**